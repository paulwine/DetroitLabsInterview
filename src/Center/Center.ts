import React from "react";
import { ForecastDay } from "../Models/ForecastDay";


export enum ActiveView {
    CurrentTemp,
    FiveDay
}

interface CenterState {
    activeView: ActiveView;
    userLocation?: {
        latitude: number;
        longitude: number;
    }; 
    userCity?: string;
    currentTemp : number;
    forecastDays: ForecastDay[];
}

const apiKey : string = '475427f62c3d25e731ed6a19c1e83f92';
const apiEndpoint : string = 'https://api.openweathermap.org/data/2.5/';
const iconEndpoint = "http://openweathermap.org/img/w/";

export class Center {
    private static instance: Center;
    public state : CenterState;

    private constructor() { 
        this.state = {
            activeView: ActiveView.CurrentTemp,
            currentTemp: 0,
            forecastDays: []
        }
    }

    public static center(): Center {
        if (!Center.instance) {
            Center.instance = new Center();
        }
        return Center.instance;
    }

    // ================== Update Functions ============================
    private listeners: React.Component[] = [];

    public static addListener(listener: React.Component) {
        Center.center().listeners.push(listener);
    }

    public static removeListener(listener: React.Component) {
        const index = Center.center().listeners.indexOf(listener);
        if (index > -1) {
            Center.center().listeners.splice(index, 1);
        }
    }

    public static update(): void {
        Center.center().listeners.forEach(listener => listener.setState({}));
    }

    //================= UI Functions =====================================

    public static navigate(view: ActiveView) : void {
        Center.center().state.activeView = view;
        Center.update();
    }

    //================== Data Functions =====================================

    public static setUserLocation() {
        navigator.geolocation.getCurrentPosition((position) => {
            let coords = {
                latitude: position.coords.latitude,
                longitude: position.coords.longitude
            }
            Center.center().state.userLocation = coords;
            Center.update();
            Center.getWeatherData();
        })

    }

    public static async getWeatherData() : Promise<void> {
        const state = Center.center().state;
        if (!state.userLocation){
            console.error("No User Location Set");
            return;
        }
        const lat = state.userLocation.latitude;
        const lon = state.userLocation.longitude;

        const currentTempEndpoint = apiEndpoint + 
        `weather?lat=${lat}&lon=${lon}&appid=${apiKey}&units=imperial`;

        const currentTempResponse = await fetch(currentTempEndpoint);
        const currentTempJSON = await currentTempResponse.json();
        const currentTemp = currentTempJSON["main"]["temp"];

        state.currentTemp = currentTemp;
        state.userCity = currentTempJSON["name"];

        const fiveDayEndpoint = apiEndpoint + 
        `forecast?lat=${lat}&lon=${lon}&appid=${apiKey}&units=imperial`;

        const fiveDayResponse = await fetch(fiveDayEndpoint);
        const fiveDayJSON = await fiveDayResponse.json();

        const list = fiveDayJSON["list"];

        let forecastDays: ForecastDay[] = [];

        list.forEach((item : any) => {
            const temperature : number = item["main"]["temp"];
            const imgUrl : string = iconEndpoint+ item["weather"][0]["icon"] + '.png';
            const date : Date = new Date(item["dt_txt"]);

            const forecastDay : ForecastDay = {
                imgUrl,
                temperature,
                date
            };
            forecastDays.push(forecastDay);
        })

        state.forecastDays = forecastDays;

        Center.update();
    }
}