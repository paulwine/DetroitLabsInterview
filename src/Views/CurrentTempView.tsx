import { Typography } from '@material-ui/core';
import React from 'react';
import { Center } from '../Center/Center';

export default class CurrentTempView extends React.Component {
    constructor(props : any){
        super(props);
    }

    componentDidMount() {
        Center.addListener(this);
    }
    componentWillUnmount() {
        Center.removeListener(this);
    }


    render(){
        const state = Center.state();
        if (!state.currentTemp){
            return (
                <div>
                    <h2>User Location Not Set</h2>
                </div>
            )
        }
        return (
        <div>
            <Typography variant={"h4"}>Current Temperature</Typography>
            <h1>{state.currentTemp}Â°F</h1>
        </div>
        )
    }
}