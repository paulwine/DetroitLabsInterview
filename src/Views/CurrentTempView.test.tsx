import 'regenerator-runtime/runtime'
import {shallow, configure } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16';
import React from 'react'
import {Center, ActiveView } from '../Center/Center'
import CurrentTempView from './CurrentTempView'

configure({ adapter: new Adapter() });

describe('CurentTempView', () =>{
    beforeAll(() => {
        Center.center().internalState = {
            userCity: undefined,
            userLocation: {
                latitude: 43.500096,
                longitude: -81.935808
            },
            forecastDays: [],
            currentTemp: 90,
            activeView: ActiveView.CurrentTemp
        }
    }),
    it('should render component to match snapshot' , async () => {
        const currentTempView = shallow(<CurrentTempView></CurrentTempView>)
        expect(currentTempView.html()).toMatchSnapshot();
    })
})