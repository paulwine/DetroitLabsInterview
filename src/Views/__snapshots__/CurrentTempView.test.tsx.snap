// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurentTempView should render component to match snapshot 1`] = `"<div><h4 class=\\"MuiTypography-root MuiTypography-h4\\">Current Temperature</h4><h1>90Â°F</h1></div>"`;
